File: Computer_Networks/CRC.java
package weblab;

class CRC {
  static int size(long input){
    int output = 0;
    while(input>=1){
      output++;
      input=input/2;
    }
    return output;
  }
  

  /**
   * Calculates the CRC check value
   *
   * @param bitSequence       The input bit sequence
   * @param inputLength       The length of the input bit sequence (including possible leading zeros)
   * @param generatorSequence The generator bit sequence
   * @return The CRC check value
   */
  static long calculate(long bitSequence, int inputLength, long generatorSequence) {

    int size1=size(generatorSequence);
    size1--;
    bitSequence = bitSequence<<size1;

    while(size(bitSequence)>size1){
      long chislochislo=generatorSequence<<(size(bitSequence)-size1-1);
      bitSequence = bitSequence ^ chislochislo;
    }
    return bitSequence;
  }

  /**
   * Checks the correctness of the bit sequence.
   *
   * @param bitSequence       The CRC bit sequence excluding the CRC check value
   * @param inputLength       The length of the input bit sequence (including possible leading zeros)
   * @param generatorSequence The generator bit sequence used
   * @param checkSequence     The CRC check value to check against
   * @return true if the sequence is valid, false otherwise
   */
  static boolean check(long bitSequence, int inputLength, long generatorSequence, long checkSequence) {
    
    if(calculate(bitSequence,inputLength, generatorSequence)!=checkSequence) return false;
    return true;
  }
}

File: Computer_Networks/CRC.py
from library import bit_count


class CRC:

    @staticmethod
    def calculate(bit_sequence: int, input_length: int, generator_sequence: int) -> int:
        size = generator_sequence.bit_length();
        bit_sequence=bit_sequence<<(size-1);
        while bit_sequence.bit_length()>(size-1):
            intermediate = generator_sequence<<(bit_sequence.bit_length()-size)
            bit_sequence=bit_sequence^intermediate
        print(bit_sequence)
        return bit_sequence


    @staticmethod
    def check(bit_sequence: int, input_length: int, generator_sequence: int, check_sequence: int) -> bool:
        size = generator_sequence.bit_length();
        bit_sequence=bit_sequence<<(size-1);
        while bit_sequence.bit_length()>(size-1):
            intermediate = generator_sequence<<(bit_sequence.bit_length()-size)
            bit_sequence=bit_sequence^intermediate
        if bit_sequence==check_sequence:
            return True
        
        return False
